sync/l1verifier.go

```diff
@@ -0,0 +1,117 @@
+package sync
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/NethermindEth/juno/blockchain"
+	"github.com/NethermindEth/juno/clients"
+	"github.com/NethermindEth/juno/utils"
+	"github.com/sourcegraph/conc/stream"
+)
+
+type ErrL1VerificationFailed struct {
+	Height uint64
+	Err    error
+}
+
+func (e ErrL1VerificationFailed) Error() string {
+	return fmt.Sprintf("L1 verification failed on block #%d with %s", e.Height, e.Err.Error())
+}
+
+// L1Verifier is a background worker that continously tries to detect the highest height that was verified
+// on L1.
+type L1Verifier struct {
+	Blockchain *blockchain.Blockchain
+	Gateway    *clients.GatewayClient
+	Wait       time.Time
+	Height     uint64
+	log        utils.SimpleLogger
+}
+
+// NewL1Verifier creates a new L1Verifier
+func NewL1Verifier(bc *blockchain.Blockchain, gateway *clients.GatewayClient, log utils.SimpleLogger) *L1Verifier {
+	return &L1Verifier{
+		Blockchain: bc,
+		Gateway:    gateway,
+		log:        log,
+	}
+}
+
+// Run starts the L1Verifier.
+func (v *L1Verifier) Run(syncCtx context.Context, verifier *stream.Stream, nextVerifiedHeight uint64) stream.Callback {
+	blockHeight := uint64(0)
+	if h, err := v.Blockchain.Height(); err == nil {
+		blockHeight = h + 1
+	}
+
+	errChan := make(chan ErrL1VerificationFailed)
+
+	verifierCtx := context.Background()
+	streamCtx, streamCancel := context.WithCancel(verifierCtx)
+	// Wait for the block to be synced before trying to verify it.
+	if nextVerifiedHeight < blockHeight {
+		select {
+		case err := <-errChan:
+			streamCancel()
+			nextVerifiedHeight = err.Height
+			v.log.Warnw("Retrying L1 verification", "height", nextVerifiedHeight, "err", err.Err)
+		case <-syncCtx.Done():
+			verifier.Wait()
+			if errors.Is(syncCtx.Err(), context.Canceled) {
+				return func() {}
+			} else {
+				return func() { syncCtx.Err() }
+			}
+		default:
+			curHeight := nextVerifiedHeight
+			return func() {
+				verifier.Go(
+					func() stream.Callback {
+						return v.verifier(streamCtx, curHeight, errChan)
+					})
+			}
+		}
+	}
+	// }
+	return func() {}
+}
+
+func (v *L1Verifier) verifier(ctx context.Context, nextVerifiedHeight uint64, errChan chan ErrL1VerificationFailed) stream.Callback {
+	return func() {
```

Comment:
this provides no concurrency, callbacks are executed synchronously

Reply:
That was intentional because `StoreL1ChainHeight` needs to be synchronous just like we do with block storage.What about just moving `GetBlock` so it will be called before the callback?

Reply:
You shouldn't need to use any streams here. You only need a single goroutine to make this work, not a pool of workers.

Reply:
Got it. Thanks